contracts/protocols/GnosisAuction.sol

```diff
@@ -1,4 +1,114 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.7.3;
+pragma experimental ABIEncoderV2;
 
-contract GnosisAuction {}
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import {IGnosisAuction} from "../interfaces/IGnosisAuction.sol";
+import {IOtoken} from "../interfaces/GammaInterface.sol";
+import {IOptionsPremiumPricer} from "../interfaces/IRibbon.sol";
+import "hardhat/console.sol";
+
+library GnosisAuction {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    event InitiateGnosisAuction(
+        address auctioningToken,
+        address biddingToken,
+        uint256 auctionCounter,
+        address manager
+    );
+
+    struct AuctionDetails {
+        address oTokenAddress;
+        address asset;
+        address underlying;
+        address manager;
+        uint256 premiumDiscount;
+        uint256 duration;
+    }
+
+    function startAuction(
+        address gnosisEasyAuction,
+        address optionsPremiumPricer,
+        AuctionDetails memory auctionDetails
+    ) internal {
+        (uint96 optionPremium, uint96 oTokenSellAmount) =
+            setupAuctionParameters(
+                auctionDetails.oTokenAddress,
+                gnosisEasyAuction,
+                auctionDetails.underlying,
+                optionsPremiumPricer,
+                auctionDetails.premiumDiscount
+            );
+
+        IERC20(auctionDetails.oTokenAddress).safeApprove(gnosisEasyAuction, 0);
+        IERC20(auctionDetails.oTokenAddress).safeApprove(
+            gnosisEasyAuction,
+            IERC20(auctionDetails.oTokenAddress).balanceOf(address(this))
+        );
```

Comment:
Since it's a new otoken, it's safe to assume the approval for the auction contract should be `0`. One way to save gas is to check if the approval != 0 before doing the double approve

```solidity
if (IERC20(auctionDetails.oTokenAddress).approval(address(this), gnosisEasyAuction) > 0) {
  IERC20(auctionDetails.oTokenAddress).safeApprove(gnosisEasyAuction, 0);
}
```

Reply:
👍